
## Module 1 - Introduction to NumPy


## 1. What is NumPy?

### Definition

NumPy stands for Numerical Python.

It is a Python library used for numerical calculations and working with arrays.

NumPy provides fast and efficient operations for handling numerical data.

### Real-world Example

NumPy can be compared to a calculator that can work with many numbers at the same time.

### Business Example

A company can use NumPy to calculate sales totals, averages, and other numerical values from large datasets.

### AI/ML Example

NumPy is used in AI/ML for numerical calculations, data processing, arrays, and matrix operations.

### Python Implementation

In [4]:
import numpy as np

data = np.array([10, 20, 30, 40, 50])

print(data)

[10 20 30 40 50]


###  Explanation

The code creates a NumPy array using `np.array()`.

The output displays the values stored in the NumPy array.

NumPy arrays are useful for performing numerical operations efficiently.

## 2. Why NumPy?

NumPy is used because it makes numerical calculations easier and faster.

It can work with many numbers at the same time using arrays.

### Best Practices

- Use NumPy arrays for numerical data.
- Use NumPy functions for calculations.
- Avoid unnecessary loops when NumPy operations can be used.

### Common Mistakes

- Forgetting to import NumPy.
- Using a Python list when an array is more suitable.
- Using incorrect NumPy functions.

## 3. NumPy Arrays (ndarray)

### Definition

A NumPy array is a collection of values stored in an organized structure.

NumPy arrays are called `ndarray`, which means N-dimensional array.

They can store data in one or more dimensions.

### Real-world Example

A list of daily temperatures can be stored in a NumPy array.

For example:

25, 27, 29, 30, 28

### Business Example

A company can store monthly sales data in a NumPy array and perform calculations on all months together.

### AI/ML Example

Machine Learning models work with numerical data, so NumPy arrays are commonly used to store and process features and values.

### Python Implementation

In [5]:
import numpy as np

data = np.array([25, 27, 29, 30, 28])

print(data)
print(type(data))

[25 27 29 30 28]
<class 'numpy.ndarray'>


###  Explanation

The `np.array()` function creates a NumPy array.

The `type()` function shows that the array is a NumPy `ndarray`.

The array stores all the values in a single structure.

## 4. Creating Arrays

### Definition

NumPy provides different ways to create arrays.

We can create arrays from Python lists, ranges of numbers, or arrays filled with zeros and ones.

### Real-world Example

If we want to store the numbers from 1 to 5, we can create a NumPy array instead of writing each value separately.

### Python Implementation

In [6]:
import numpy as np

a = np.array([1, 2, 3, 4, 5])
b = np.arange(1, 6)
c = np.zeros(3)

print(a)
print(b)
print(c)

[1 2 3 4 5]
[1 2 3 4 5]
[0. 0. 0.]


### Explanation

`np.array()` creates an array from the given values.

`np.arange()` creates an array with numbers in a given range.

`np.zeros()` creates an array where all values are zero.

### Best Practices

- Choose the array creation method based on the requirement.
- Use `arange()` when working with a sequence of numbers.
- Use `zeros()` when an array needs to be initialized with zeros.

### Common Mistakes

- Using the wrong range with `arange()`.
- Forgetting that the stop value in `arange()` is not included.
- Creating an array with an incorrect size.

## 5. Dimensions

### Definition

The dimension of a NumPy array tells us how many levels or axes the array has.

A 1D array has one axis, while a 2D array has rows and columns.

### Real-world Example

A list of student marks can be represented using a 1D array.

A table containing marks for multiple students and subjects can be represented using a 2D array.

### AI/ML Example

Machine Learning datasets are often represented as 2D arrays, where rows represent data samples and columns represent features.

### Python Implementation

In [8]:
import numpy as np

one_d = np.array([10, 20, 30])
two_d = np.array([[10, 20], [30, 40]])

print("1D:", one_d.ndim)
print("2D:", two_d.ndim)

1D: 1
2D: 2


###  Explanation

`ndim` returns the number of dimensions of an array.

The first array has 1 dimension, and the second array has 2 dimensions.

## 6. Shape

### Definition

Shape tells us the size of each dimension in a NumPy array.

For a 2D array, shape is represented as `(rows, columns)`.

### Python Implementation

In [9]:
import numpy as np

data = np.array([
    [10, 20],
    [30, 40],
    [50, 60]
])

print("Shape:", data.shape)

Shape: (3, 2)


### Explanation

The shape of the array is `(3, 2)`.

This means the array has 3 rows and 2 columns.

## 7. Size

### Definition

Size tells us the total number of elements present in a NumPy array.

It counts all values in the array, regardless of how many dimensions it has.

### Real-world Example

If an array contains the marks of 10 students, its size is 10.

If a table has 5 rows and 3 columns, its size is 15.

### Python Implementation

In [10]:
import numpy as np

data = np.array([
    [10, 20],
    [30, 40],
    [50, 60]
])

print("Size:", data.size)

Size: 6


###  Explanation

The array contains 3 rows and 2 columns.

Therefore, the total number of elements is 3 × 2 = 6.

The `.size` attribute returns 6.

### Best Practices

- Use `.size` to find the total number of elements.
- Check the size when working with large datasets.
- Use `.shape` when you need row and column information.

### Common Mistakes

- Confusing size with shape.
- Assuming `.size` returns the number of rows.
- Forgetting that all elements across dimensions are counted.

## 8. Data Types

### Definition

Data type tells us what kind of values are stored in a NumPy array.

NumPy supports data types such as integers, floating-point numbers, and strings.

### Real-world Example

Student ages can be stored as integers, while student percentages can be stored as floating-point numbers.

### Business Example

A company can store product quantities as integers and product prices as floating-point values.

### AI/ML Example

Choosing a suitable data type can help reduce memory usage and improve the efficiency of numerical data processing.

### Python Implementation

In [11]:
import numpy as np

data = np.array([10, 20, 30, 40])

print("Data:", data)
print("Data Type:", data.dtype)

Data: [10 20 30 40]
Data Type: int64


### Explanation

The code creates a NumPy array containing integer values.

The `.dtype` attribute shows the data type of the values stored in the array.

Here, the values are stored as an integer type.

## 9. Indexing

### Definition

Indexing is used to access a specific element from a NumPy array.

NumPy uses zero-based indexing, so the first element has index 0.

### Python Implementation

In [13]:
import numpy as np

data = np.array([10, 20, 30, 40, 50])

print("First value:", data[0])
print("Third value:", data[2])

First value: 10
Third value: 30


###  Explanation

`data[0]` accesses the first value in the array.

`data[2]` accesses the third value because NumPy uses zero-based indexing.

The index starts from 0, not 1.

### Best Practices

- Remember that indexing starts from 0.
- Use valid index positions.
- Check the array size before accessing an index.

### Common Mistakes

- Starting the index from 1.
- Using an index that does not exist.
- Confusing the index position with the actual value.

## 10. Slicing

### Definition

Slicing is used to select a range of elements from a NumPy array.

The basic syntax is `array[start:stop]`.

The start index is included, but the stop index is not included.

### Real-world Example

If an array contains daily sales values, slicing can be used to select sales from a particular group of days.

### Business Example

A company can use slicing to select sales data for a specific period from a larger dataset.

### AI/ML Example

Slicing can be used to select specific values or features from numerical data during data preprocessing.

### Python Implementation

In [14]:
import numpy as np

data = np.array([10, 20, 30, 40, 50])

print(data[1:4])

[20 30 40]


###  Explanation

`data[1:4]` selects values from index 1 up to index 4.

The value at index 4 is not included.

The output is `[20 30 40]`.

## 11. Reshaping

### Definition

Reshaping is the process of changing the structure of a NumPy array without changing its values.

It is useful when we need to arrange data into a different number of rows and columns.

### Python Implementation

In [16]:
import numpy as np

data = np.array([1, 2, 3, 4, 5, 6])

reshaped = data.reshape(2, 3)

print(reshaped)

[[1 2 3]
 [4 5 6]]


###  Explanation

`reshape(2, 3)` changes the array into 2 rows and 3 columns.

The values remain the same, but their arrangement is changed.

## 12. Flatten

### Definition

Flatten is used to convert a multi-dimensional NumPy array into a one-dimensional array.

It keeps all the values but removes the row and column structure.

### Python Implementation

In [17]:
import numpy as np

data = np.array([[1, 2], [3, 4], [5, 6]])

flat = data.flatten()

print(flat)

[1 2 3 4 5 6]


###  Explanation

`flatten()` converts the 2D array into a 1D array.

The values remain the same, but the row and column structure is removed.

The output is `[1 2 3 4 5 6]`.

## 13. Copy vs View

### Definition

A copy creates a separate array with its own data.

A view shares the same underlying data with the original array.

Because a view shares the data, changes made to the view can also affect the original array.

### Real-world Example

A copy is like making a separate photocopy of a document.

A view is like looking at the same original document from another place.

### AI/ML Example

Copy and view are important when preparing large datasets because creating unnecessary copies can increase memory usage.

### Python Implementation

In [18]:
import numpy as np

data = np.array([10, 20, 30, 40])

copy_data = data.copy()
view_data = data.view()

copy_data[0] = 100
view_data[1] = 200

print("Original:", data)
print("Copy:", copy_data)
print("View:", view_data)

Original: [ 10 200  30  40]
Copy: [100  20  30  40]
View: [ 10 200  30  40]


###  Explanation

The copy is independent, so changing `copy_data` does not change the original array.

The view shares data with the original array, so changing `view_data` also changes the original array.

This shows the main difference between a copy and a view.

## 14. Broadcasting

### Definition

Broadcasting allows NumPy to perform operations between arrays with different shapes.

NumPy automatically adjusts the smaller array so that the operation can be performed.

### Real-world Example

Suppose we have prices for several products and want to add the same tax amount to all products.

Broadcasting allows us to add the tax value to all prices at once.

### Python Implementation


In [19]:
import numpy as np

prices = np.array([100, 200, 300])
tax = 10

result = prices + tax

print(result)

[110 210 310]


###  Explanation

The value `10` is automatically applied to every element of the `prices` array.

The output is `[110 210 310]`.

This automatic handling of different shapes is called broadcasting.

## 15. Array Operations

### Definition

Array operations are mathematical operations performed directly on NumPy arrays.

NumPy allows us to add, subtract, multiply, and divide array values without using loops.

### Real-world Example

If we have the prices of products and want to increase all prices by 10, we can perform the operation on the complete array.

### Python Implementation

In [21]:
import numpy as np

a = np.array([10, 20, 30])
b = np.array([2, 4, 6])

print("Addition:", a + b)
print("Subtraction:", a - b)
print("Multiplication:", a * b)
print("Division:", a / b)

Addition: [12 24 36]
Subtraction: [ 8 16 24]
Multiplication: [ 20  80 180]
Division: [5. 5. 5.]


### Explanation

The code performs addition, subtraction, multiplication, and division on two NumPy arrays.

The operations are performed element by element.

NumPy allows these calculations without writing separate loops.

## 16. Universal Functions (ufuncs)

### Definition

Universal functions, or ufuncs, are NumPy functions that perform operations on each element of an array.

They work efficiently on entire arrays without requiring explicit loops.

### Real-world Example

If we have temperatures in Celsius, we can use a ufunc to perform calculations on all temperature values at once.

### AI/ML Example

Ufuncs are used in AI/ML for mathematical transformations and numerical calculations during data processing.

### Python Implementation

In [22]:
import numpy as np

data = np.array([1, 4, 9, 16])

result = np.sqrt(data)

print(result)

[1. 2. 3. 4.]


### Explanation

`np.sqrt()` is a NumPy universal function that calculates the square root of each value.

The operation is applied to every element in the array.

The output contains the square roots of the original values.

## Advantages

- Fast numerical calculations.
- Efficient handling of arrays.
- Supports mathematical and statistical operations.
- Uses memory efficiently for numerical data.
- Useful for Data Science, AI, and Machine Learning.

## Limitations

- Mainly designed for numerical data.
- Large arrays can require more memory.
- Requires understanding of dimensions and array operations.
- Not suitable for complex table-based data management.
- Some operations may require knowledge of NumPy syntax.

## Best Practices

- Use NumPy arrays for numerical data and calculations.
- Check the shape, size, and data type before performing operations.
- Use vectorized operations instead of unnecessary loops.
- Use `copy()` when the original data should not be changed.
- Make sure array shapes are compatible when using broadcasting.

## Common Mistakes

- Confusing NumPy arrays with Python lists.
- Using incorrect indexing or slicing positions.
- Trying to reshape an array into an incompatible shape.
- Confusing a copy with a view.
- Using incompatible array shapes during operations.

## Interview Questions and Answers

### 1. What is NumPy?

NumPy is a Python library used for numerical computing.
It provides arrays and functions for mathematical operations.
It is widely used in Data Science, AI, and Machine Learning.

### 2. Why is NumPy used?

NumPy makes numerical calculations faster and easier.
It can perform operations on many values at the same time.
It is useful for handling large numerical datasets.

### 3. What is a NumPy array?

A NumPy array is a collection of values stored in an organized structure.
It is represented using the `ndarray` object.
It can have one or multiple dimensions.

### 4. What is the difference between a Python list and NumPy array?

A Python list is a general-purpose collection of values.
A NumPy array is designed mainly for numerical operations.
NumPy arrays are generally more efficient for numerical calculations.

### 5. What is the difference between shape and size?

Shape tells us the dimensions of an array, such as rows and columns.
Size tells us the total number of elements in the array.
For example, an array with shape `(2, 3)` has a size of 6.

### 6. What is indexing in NumPy?

Indexing is used to access a specific element from an array.
NumPy uses zero-based indexing.
This means the first element has index 0.

### 7. What is slicing in NumPy?

Slicing is used to select a range of elements from an array.
The basic syntax is `array[start:stop]`.
The start position is included, but the stop position is excluded.

### 8. What is broadcasting?

Broadcasting allows NumPy to perform operations on arrays with compatible shapes.
It can apply a single value to multiple elements automatically.
This makes numerical operations simple and efficient.

### 9. What is the difference between copy and view?

A copy creates a separate array with independent data.
A view shares the same underlying data with the original array.
Changes to a view can also affect the original array.

### 10. What are universal functions (ufuncs)?

Universal functions are NumPy functions that operate on array elements.
Examples include `np.sqrt()`, `np.exp()`, and `np.sin()`.
They perform calculations efficiently without requiring explicit loops.

## Self Reflection

I learned the basic concepts of NumPy and how NumPy arrays work.

I practiced creating, indexing, slicing, reshaping, and performing operations on arrays.

I also understood how NumPy is useful for numerical calculations and AI/ML data processing.